spec.html

```diff
@@ -11059,6 +11060,44 @@ <h1>Examples of Automatic Semicolon Insertion</h1>
+
+    <p>The term &ldquo;<dfn>automatic semicolon insertion hazard</dfn>&rdquo; refers, to a place where a semicolon may or may not be inserted depending on preceding source text according to the rules above. These places are considered hazards if changes outside of an existing source text grammar production could change the grammar production of the existing source text depending on automatic semicolon insertion. The rest of this section describes a number of automatic semicolon insertion hazards in this version of ECMAScript.</p>
+    <emu-clause id="sec-asi-hazards-in-statement-lists">
+      <h1>Automatic Semicolon Insertion Hazards in Statement Lists</h1>
+      <p>In a |StatementList|, many |StatementListItem|s end in semicolons, which may be omitted using automatic semicolon insertion. As a consequence of the rules above, at the end of a line ending an expression, a semicolon is required if the following line begins with any of the following:
+      <ul>
+        <li><strong>An opening parenthesis (<code>(</code>)</strong>. Without a semicolon, the two lines together are treated as a |CallExpression|.</li>
+        <li><strong>An opening square bracket (<code>[</code>)</strong>. Without a semicolon, the two lines together are treated as property access, rather than an |ArrayLiteral| or |ArrayAssignmentPattern|.</li>
+        <li><strong>A template string (<code>`</code>)</strong>. Without a semicolon, the two lines together are interpreted as a tagged Template (<emu-xref href="#sec-tagged-templates"></emu-xref>), with the previous expression as the |MemberExpression|.</li>
+        <li><strong>Unary <code>+</code> or <code>-</code></strong>. Without a semicolon, the two lines together are interpreted as a usage of the corresponding binary operator.</li>
+        <li><strong>A RegExp literal</strong>. Without a semicolon, the two lines together may be parsed instead as the `/` |MultiplicativeOperator|, for example if the RegExp has flags.</li>
+      </ul>
+    </emu-clause>
+  </emu-clause>
+  <emu-clause id="sec-no-lineterminator-here-automatic-semicolon-insertion">
+    <h1>Automatic Semicolon Insertion and &ldquo;[no |LineTerminator| here]&rdquo;</h1>
+    <em>This section is non-normative.</em>
+    <p>ECMAScript contains grammar productions which include &ldquo;[no |LineTerminator| here]&rdquo;. These productions are sometimes a means to have optional operands in the grammar. Introducing a |LineTerminator| in these locations would change the grammar production of a source text by using the grammar production without the optional operand. These locations are not considered hazardous because additive changes would need to alter source text within a grammar production that is subject to automatic semicolon insertion in order to change the grammar production.</p>
```

Comment:
Thank you for adding the section about `[no LineTerminetor here]`. However, I think "not considered hazardous" is not proper. This explanation looks to be written from the perspective of semicolon-less style. From the perspective of people which write semicolons always, "These productions are sometimes a means to have optional operands in the grammar" is weird because not sure why `return\nExpression;` is not allowed. Splitting `return\nExpression;` by ASI is still a hazard.

Reply:
@mysticatea can you suggest a different phrasing to this perhaps? the difference is that source text outside of existing productions can affect other productions for the hazards. The `[no LineTerminator here]` situations do not change behavior if source text surrounding them changes, and require changes to the already used production to change the behavior. Specifically I want to preserve that those cases are not going to add syntax errors/change productions from their surroundings changing.

Reply:
@bmeck In my think, ASI hazards exist two kinds.

1. **Unexpected joining two statements.** This is caused by the wrong assumption that line breaks separate statements consistently. I think this is described enough. (But I think ASI doesn't change behavior in this case. Just people expected ASI to change behavior but not in fact.)
2. **Unexpected splitting one statement.** This is caused by the wrong assumption that semicolons are the only delimiter for statements. The `[no LineTerminator here]` situations should be syntax error if `;` is lacking and ASI doesn't exist. (Or it will connect to the next line if the `[no LineTerminator here]` doesn't exist.) However, ASI *changes* the behavior as hiding the syntax error and working in an unexpected way. For example, `return\n42;` case, a semicolon is lacking but ASI hides the syntax error (and then returning undefined can cause another `TypeError`).

I think that the two kinds should be described in the same level rather than "one is hazards but another is not hazards". Both are ASI Hazards.

I apology, It's difficult for me to suggest formal phrase because of my poor English skill.

Reply:
@mysticatea 

1. requires no action be performed to edit statements, 2 independent statements can be joined without editting either. This can even result in syntax errors.

```sh
echo 'let a = 0' > a
echo '/b/.test()' > b
cat a b > c
```

For example has a situation where 2 statements create a syntax error without any changes to the statements.

```sh
echo 'let a = 0' > a
echo '+(()=>{})()' > b
cat a b > c
```

Likewise alters the meanings of the source texts without either individual text be split.

2. requires explicit action be performed to change from one statement to two and is not subject to adjacent code.

```sh
echo 'return' > a
echo '{}' > b
cat a b > c
```

`a` is not affected by `b`.

In order for there to be a problem a statement needs to be altered internally (such as by adding a newline).

Reply:
@bmeck I understand that you are trying to show how case (1.) is hazardous and case (2.) is not. But this is according to *your* definition of “hazard,” which I and @mysticatea see to be *too narrow.*

I and @mysticatea saw both (1.) and (2.) as hazardous, but that these are 2 kinds of hazards.

1. Let’s call it **“line concatenation hazard.”** This one is hazardous because 2 lines of statement, when concatenated, resulted in a single statement with unexpected behavior. This can happen when a developer put code from 2 statements next to each other.

2. Let’s can call it **“line break hazard.”** This one is also hazardous because most C-like languages are whitespace insensitive, and users familiar with C/C++/Java/C#/PHP/… may expect that adding a whitespace (space, tab, linebreak) should not alter the meaning of code. However, [no LineTerminator here] cases break that assumption. This can happen when a developer adjusts the formatting of code, e.g. breaking a long line for readability.

In both cases (1.) and cases (2.), I showed that:

- **A developer did something seemingly innocent to the code:**
    1. Putting 2 statements together
    2. Adjust formatting of code

- **The developer expected a result similar to a language they are familiar with:**
    1. statement1 should run, followed by statement2 (like in Ruby/Python/Go/…)
    2. Meaning of code should not be altered (like in C/Java/PHP/…)

- **But then something different happened:**
    1. Code statement1 and statement2 joined into a statement with entirely different meaning, leading to unexpected behavior.
    2. Code is split into 2 statements, altering the meaning of code, leading to unexpected behavior.

For me, these three points can be a precise-enough definition of the word “hazard,” without being too narrow to a single viewpoint. As shown, both cases (1.) and (2.) satisfies these 3 ingredients. Therefore, I consider them hazards.

Reply:
@dtinth While both situations may subjectively be called confusing or may not. The choice of calling one a hazard is a violation of locality that can be used as an attack vector in addition to this. The first form of this PR did not include the `[no LineTerminator here]` because it was only subjectively confusing, however comments on the PR have led to the addition of a list of such situations being included.

While the points above are a good description of the subjective side of things, they are not about risk assessment. JS engines and tools often can create source text based upon other source text. `Function`, `eval`, `blob:`/`text:` URLs, etc. can all be used for runtime code generation in environments. The insertion hazard has a risk that code being used to template source text generation is not isolated to a specific locality. This same hazard also applies to various tools doing code generation of source texts. For tooling and templating neither of these cases apply to the `[no LineTerminator here]` locations.

I'm open to renaming the term `hazard` but it is strictly a risk assessment concern that does not apply to the `[no LineTerminator here]` situations. If another form of risk is found we can happily document/add whatever text makes sense just like we added the list of `[no LineTerminator here]` list even without a violation of locality.

Reply:
@bmeck I learned your definition. However, I'm not sure why you address bunder tools (the tool which concatenates multiple codes to one) as a special one. There are several kinds of tools, concatenation problems are related to the tool like bundlers, and `[no LineTerminator here]` situations are related to the tool like formatters. Why is the bundler-like tool special?

Reply:
@mysticatea Bundlers are just the most common tools outside programmers themselves that mutate source code :-). The main note here is that the `[no LineTerminator here]` does not suffer a locality problem because the source text manipulation must be done within a Grammar Production that is not separated by a semicolon.

Reply:
Instead of a Grammar Production *adjacent* to another.

Reply:
@bmeck Thank you for patience in this discussion.

> Bundlers are just the most common tools outside programmers themselves that mutate source code

I'm not sure I agree with your assumption. For example, most editors/IDEs embed formatters and many people use it, and many many packages which don't use any bundler exist in npm repository. At least, `cat` is not a common tool to concatenate js codes. Bundler tools are not special in tools which modify source code texts.

> The main note here is that ...

I'm not sure why you think that "inside or outside of Grammar Production" is an important difference.

- Common tools can modify both. It can add a line break then split one Production to two Productions. It can remove a semicolon then join two Productions to one Production. The result (the change of the number of Productions) is same.
- I can't find significant difference between "To use `eval(codeList.join(""))` instead of `eval(codeList.join(";"))`" and "To use `return\nobj` instead of `return obj`". Both are caused by wrong assumptions of the language spec (https://github.com/tc39/ecma262/pull/1062#discussion_r161385712).
- Both kinds can increase. People must update their knowledge or use linter-like tools in order to prevent problems. For example, line break after `async` keyword causes a problem, lacking semicolon before `ArrayPattern` in LHS causes a problem. There is no difference -- new syntax can introduce new problem of ASI.

I afraid that this section will lead people to underestimate one surface of ASI problems.

Reply:
> I'm not sure why you think that "inside or outside of Grammar Production" is an important difference

Because it is a violation of locality. I have not seen in the past any indication of within grammar production discussions of ASI. Hence things like `yield`, [the debugger operand proposal](https://github.com/bmeck/proposal-debugger-operands), an talk of making the throw operand optional [like CSharp](https://blogs.msdn.microsoft.com/jmstall/2007/02/15/throw-e-vs-throw/), etc. Since these alterations have not and are not expected to violate locality I don't think that the committee would be giving heavy weight to their additions on a grammar level.

Both list will grow, both may or may not be seen as confusing, but only one has a locality problem which has bad implications regarding what grammar productions are able to be considered safely isolated. The inability to have this isolation is not an opinion and the difference in stated risk is to convey that. I would need some risk of locality violation or likewise from the `[no LineTerminator here]` to declare them to be a hazard. Doing so would likely remove any operators with optional operands from the future spec I might suspect if we found one and I don't believe there is one already discovered.

Reply:
@bmeck OK, I probably learned it.

There are two actors here.

- **Producer** ... people which write the spec.
- **Consumer** ... people/tools which use the spec. (E.g., JS programmers)

My feeling came from the difference of the definition of "hazards" between Producer and Consumer. This section is a notice for Consumer but described with Producer's definition. In the perspective of Consumer, both "line concatenation hazard" and "linebreak hazard" are important as ASI problems because lack of ASI knowledge causes both problems in equally. However, in the perspective of Producer, "linebreak hazard" doesn't seem problems because it's encapsulated in each syntax proposal. This mismatch looked to cause confusion. When a Consumer read the notice with Consumer's definition, I guess that they feel "this makes the different weight between two kinds of ASI problems in order to recommend to use semicolons, looks unfair", and I guess it's the reason of public reactions.

Thank you.

Reply:
Maybe it's nice if the title "Hazards of Automatic Semicolon Insertion" is rephrased. This title seemed to mislead me to "this section describes all problems of ASI". How about "Hazards of Automatic Semicolon Insertion in mutual relation of syntax" or something like?

And `[no LineTerminator here]` section can mention "the `[no LineTerminator here]` isn't a problem in mutual relation of syntax because it's encapsulated in each syntactic feature".

In my 2 cents.
Thank you.

Reply:
How about changing the title to "Adjacency Hazards of Automatic Semicolon Insertion" for what currently is the "Hazards of Automatic Semicolon Insertion"? I removed the line about `[no LineTerminator here]` not being a hazard, it wasn't strictly needed since it is in a different section.

Reply:
Sounds good to me.